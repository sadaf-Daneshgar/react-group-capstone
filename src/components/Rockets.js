import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchRocketsData, reserveRocket, cancelReservation } from '../redux/rockets/rocketsSlice';
import '../style/navbar.css';

function Rockets() {
  const { rockets, isLoading, error } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();

  const handleReserveBtn = (rocketId) => {
    if (rockets.find((rocket) => rocket.id === rocketId).reserved) {
      dispatch(cancelReservation({ canceledRocketId: rocketId }));
    } else {
      dispatch(reserveRocket({ reservedRocketId: rocketId }));
    }
  };

  useEffect(() => {
    dispatch(fetchRocketsData());
  }, [dispatch]);

  if (isLoading) {
    return <div className="loading">Loading...</div>;
  }

  if (error) {
    return <div className="loading">Something went wrong during fetching rockets</div>;
  }

  return (
    <>
      <div className="container">
        {rockets.map((rocket) => (
          <div key={rocket.id} className="rocket">
            <div className="img">
              <img src={rocket.flickr_images[0]} alt={rocket.name} />
            </div>
            <div className="main-part">
              <h3 className="rocket-title">{rocket.name}</h3>
              <p className="details">
                {rocket.reserved && (
                  <span className="reserved">
                    Reserved
                  </span>
                )}
                {rocket.description}
              </p>
              <button
                type="button"
                className={`reserve-btn ${rocket.reserved ? 'reserved-btn' : ''}`}
                onClick={() => handleReserveBtn(rocket.id)}
              >
                {rocket.reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
              </button>
            </div>
          </div>
        ))}
      </div>
    </>
  );
}

export default Rockets;
