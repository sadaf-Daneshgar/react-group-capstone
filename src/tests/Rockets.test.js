import React from 'react';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Rockets from '../components/Rockets';

const mockStore = configureStore([]);

describe('Rocket Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      rockets: {
        rockets: [
          {
            id: '1',
            name: 'Sample Rocket',
            description: 'This is a sample rocket',
            flickr_images: ['image1.jpg', 'image2.jpg'],
            reserved: false,
          },
        ],
        status: 'idle',
        error: null,
      },
    });
  });

  it('renders without errors', () => {
    const component = create( 
      <Provider store={store}>
        <Rockets />
      </Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
