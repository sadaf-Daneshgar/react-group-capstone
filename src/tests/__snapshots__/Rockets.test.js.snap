// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rocket Component renders without errors 1`] = `
<div
  className="container"
>
  <div
    className="rocket-container"
  >
    <div
      className="rocket-image"
    >
      <img
        alt="Sample Rocket"
        src={
          Array [
            "image1.jpg",
            "image2.jpg",
          ]
        }
      />
    </div>
    <div
      className="main-part"
    >
      <h3
        className="rocket-title"
      >
        Sample Rocket
      </h3>
      <p
        className="details"
      >
        This is a sample rocket
      </p>
      <button
        className="reserve-btn"
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;
